### PR TITLE
Feature/printing labels #1725 and #1747

### PR DESCRIPTION
**Pull request by omelao**
Finished #1725 
I had some troubles with this task, discovering about core_fields and trying to sort by it contents. @straight-shoota  helped me a lot. Thank you so much! 😁
@diegoliberman  asked to move the replicate column to the concentration side in CSV.

This PR will continue with #1747 coming soon.

### CODE REVIEW

File: app/controllers/boxes_controller.rb
Comment (ysbaddaden) [inline app/controllers/boxes_controller.rb:126]: Checking a boolean is cheaper than running a SQL query, so I would test the other way around (it will only check the sender if the box is blinded):

```suggestion
    if !@box.blinded || is_sender
```

File: app/controllers/boxes_controller.rb
Comment (ysbaddaden) [inline app/controllers/boxes_controller.rb:144]: There is no `else` to authorize that the user actually has the READ_BOX for either the sender or receiver institution. We can probably do it by reproducing what `authorize_resource` is doing under the hood (I know @straight-shoota did).

Don't worry, I'll deal with that case before merging.

**Review by ysbaddaden — COMMENTED** [review]
It's a complex feature, and I wish we wouldn't have to implement it.

I let you finish the pull request, then I'll try to tweak some things. I still left a few comments of things that you can improve (+ a permission check that we musn't forget about).

**Comment by omelao** [comment]
@ysbaddaden I changed a lot the parts you have mentioned, now it's really finished. But a suggestion that still made sense was to change the order of `is_sender` and `blinded`, so I did it.

**Comment by omelao** [comment]
Some Transfer Packages Specs are conflicting, I'm already fixing it. So....wait for it to review please.

**Comment by omelao** [comment]
I identified some pendings in the tests that happen even in the main branch.
image

File: app/controllers/boxes_controller.rb
Comment (straight-shoota) [reply app/controllers/boxes_controller.rb:117]: There might be a deeper authorization problem here, though.

Consider that a box that has been transferred multiple times:

* Institution I creates box B
* Institution I transfers box B to institution J
* Institution J transfers box B to institution K

Should institution I still be able to access the inventory of box B, even after it has transferred to J and then K?

/cc @diegoliberman

File: app/controllers/boxes_controller.rb
Comment (straight-shoota) [inline app/controllers/boxes_controller.rb:32]: Multiple calls to `is_sender` result in multiple data base queries. The result should be cached in a local variable or instance variable and re-used.

File: app/controllers/boxes_controller.rb
Comment (omelao) [reply app/controllers/boxes_controller.rb:117]: This ticket changed over the time I was doing. The transfer package was needed considering the option to send a box multiple times, that's why I needed this parameter. But at the end, we decided to don't permit this anymore. So yes, probably, we don't need this logic anymore.

File: app/controllers/boxes_controller.rb
Comment (omelao) [reply app/controllers/boxes_controller.rb:140]: Thank's. Didn't know about `any` and that ActiveRecord figures out the pk.

File: app/controllers/transfer_packages_controller.rb
Comment (omelao) [reply app/controllers/transfer_packages_controller.rb:87]: I tried it, but it returns an error. I will put here what happened, but was some conflict with other filters, because of the samples count, this query is grouped, or something like this.

Comment (omelao) [reply app/controllers/transfer_packages_controller.rb:87]: The only way I managed to do it was this, as it is a query has a `limit 5` and almost always the user will paste a hash Id, I imagine that is not a big problem to make an additional query on this specific case.

File: app/models/box.rb
Comment (omelao) [reply app/models/box.rb:92]: I think this is solved in the ticket that did this changes, you already said this on that ticket. But I will check.

File: app/assets/stylesheets/_box_preview.scss
Comment (omelao) [reply app/assets/stylesheets/_box_preview.scss:6]: Yeah, the fixed columns has very specific contents, and I got some troubles resizing the page. But I will try your way.

File: app/controllers/boxes_controller.rb
Comment (omelao) [reply app/controllers/boxes_controller.rb:120]: > Relying on rescue as a fall back is quite inefficient. It would be better to avoid raising `RecordNotFound` in the first place! We can use the non-raising alternative to `find(id)`, which is `where(id: id).take`. If the value of that expression is `nil`, fall back to the other query.

I'm not sure if I understood....you mean like this?
```
@box = Box.where(institution: @navigation_context.institution, id: params.fetch(:id))
    if @box == nil
      @boxes = TransferPackage
        .within( @navigation_context.institution )
        .boxes
      @box = @boxes.find( params.fetch( :id ) )
    end
```

Comment (straight-shoota) [reply app/controllers/boxes_controller.rb:120]: Basically, yes. You need an `.take` on the first expression though.

File: app/controllers/transfer_packages_controller.rb
Comment (straight-shoota) [reply app/controllers/transfer_packages_controller.rb:87]: I apologize, my suggestions was full of typos 🙈 
This is correct:
```suggestion
    @boxes = @boxes.left_joins(:box_transfers).where(box_transfers: {id: nil})
```

File: app/controllers/boxes_controller.rb
Comment (omelao) [reply app/controllers/boxes_controller.rb:117]: I don't know if it's a good idea to remove transfer_package now. It's not a thing the will cause some problem, and it will be easier to make a multiple transfer functionality in the future. In this specific lines of code, I removed this param and got errors, so if I remove it, I will need to rebuild and test again the entire flow of the task. This `raise` treats the cases that the the box is not related to any institution, so I need to get the transfer. I changed this query a lot, I was doing it through BoxTransfers Models too, but this one is the sorter option.

File: app/controllers/transfer_packages_controller.rb
Comment (omelao) [reply app/controllers/transfer_packages_controller.rb:87]: Yeah, I don't know what I was doing wrong when I did this task, but I didn't get any errors this time. It works. In fact I was studying the ActiveRecord functionality, so probably I did something wrong.

```
@boxes = Box
      .within(@navigation_context.entity, @navigation_context.exclude_subsites)
      .left_joins(:box_transfers)
      .where(box_transfers: {id: nil})
      .autocomplete(uuid)
      .order("boxes.created_at DESC")
      .count_samples
      .limit(5)
```

File: app/controllers/boxes_controller.rb
Comment (omelao) [reply app/controllers/boxes_controller.rb:120]: I don't know why it doesn't work. That's why I used that `RecordNotFound` solution.
image

Comment (omelao) [reply app/controllers/boxes_controller.rb:120]: Ok, it's working

File: app/models/box.rb
Comment (omelao) [reply app/models/box.rb:92]: Yes, it's even giving conflict at this point in the code, I have already resolved the conflict

**Comment by omelao** [comment]
@diegoliberman @ysbaddaden @straight-shoota waiting for your "OK" to merge this.

File: app/controllers/boxes_controller.rb
Comment (ysbaddaden) [inline app/controllers/boxes_controller.rb:144]: In Ruby `.any?` isn't the opposite of `.empty?`. It usually involves iterating the collection for a truthy value (similar to `Array.prototype.some` in JavaScript), while `.empty?` merely checks that it's size is zero.

I'm not sure what ActiveRecord does, but the fastest method is `.exists?` (no iteration, no loading, not even a SQL count):

```suggestion
      .exists?
```

You'll may want to memoize the result, so you can repeatedly call the method. We also prefer a question mark suffix to is/has prefixes in Ruby:

```ruby
def sender?
  return @sender unless @sender.nil?
  @sender = BoxTransfer...
end
```

File: spec/controllers/boxes_controller_spec.rb
Comment (ysbaddaden) [inline spec/controllers/boxes_controller_spec.rb:168]: Nit picking:

1. you don't need to use instance variables (`@as_its` and `@sorted` are locals).
2. you can leverage map to improve readability:

```ruby
results = CSV.parse().tap(&:shift).map do |row|
  { batch_number: ... }
end
```

**Review by ysbaddaden — CHANGES_REQUESTED** [review]
Hey @omelao, great job on the feature. It's not an easy one.

I noticed a few issues to fix, but the rest is mostly nit picks to improve performance and readability!

File: app/controllers/boxes_controller.rb
Comment (omelao) [reply app/controllers/boxes_controller.rb:120]: > Where is `@boxes` defined? Did you mean `@box`? In which case:
> 
> ```ruby
> @box ||= Box.joins(...).where(...)
> ```

😅 yes, it's @box, I think it happened because I copied from a comment, talking to Johannes.

File: app/controllers/boxes_controller.rb
Comment (omelao) [reply app/controllers/boxes_controller.rb:131]: I did it spliting the print case, because we have an exception for this case. Because the order of the inventory and the printing labels must be equal to the sender. And he must have the same order to both. But he can see the information unblinded on inventory, even when it's blinded. But this situation is only for printing functions.

Comment (omelao) [reply app/controllers/boxes_controller.rb:131]: For the receiver yes, but for the sender no. On the beggining I was afraid to use the same method and change something that I can't for other screen. And it's not just the sort, but we have an exception to show blinded data for the sender in Inventory CSV. But I will merge the methods anyway. Now I'm comfortable to do this.

File: app/controllers/boxes_controller.rb
Comment (omelao) [reply app/controllers/boxes_controller.rb:144]: It was `.present?` @straight-shoota suggested `.any?`. In [this link](https://andreigridnev.com/blog/2016-04-17-empty-blank-any-exists-methods-of-ruby-on-rails-activerecord/) says that the fastest is `.exists`. So I will take your suggestion. :)

File: app/controllers/boxes_controller.rb
Comment (omelao) [reply app/controllers/boxes_controller.rb:137]: I didn't understand yet why in some cases it's reassigned and others don't. For example in your suggestion, why don't you removed the reassign of line 134?

File: app/controllers/boxes_controller.rb
Comment (straight-shoota) [reply app/controllers/boxes_controller.rb:144]: Ah, yes `exists?` is indeed better here. But that's just because of the old Rails version. Only in newer versions what I said above is true and `any?` is equivalent to `exists?`.

File: app/controllers/boxes_controller.rb
Comment (omelao) [reply app/controllers/boxes_controller.rb:122]: Because I was searching through the Transfer Packages model, and I removed the transfer_package id from all the logic. So I tried another way as I have just the box id and the institution.

Comment (straight-shoota) [reply app/controllers/boxes_controller.rb:122]: You can merge the `TransferPackage.within` scope with the Box relation:
```suggestion
        .merge(TransferPackage.within(@navigation_context.institution)).where(id: params.fetch(:id)), take
```

File: app/controllers/transfer_packages_controller.rb
Comment (omelao) [reply app/controllers/transfer_packages_controller.rb:80]: And how could I write one test for this kind of situation?

Comment (omelao) [reply app/controllers/transfer_packages_controller.rb:80]: Could I test this single method, or I would make some rule on BoxTransfer model?

File: app/controllers/boxes_controller.rb
Comment (omelao) [reply app/controllers/boxes_controller.rb:137]: Another question about this....where is the definition of `scrambled`? I just found `scrambled_samples` in Box model.

File: app/controllers/transfer_packages_controller.rb
Comment (omelao) [reply app/controllers/transfer_packages_controller.rb:80]: I will try :)

File: app/controllers/boxes_controller.rb
Comment (omelao) [reply app/controllers/boxes_controller.rb:29]: @ysbaddaden I identified one problem with this just now. I was testing sending the boxes from a user that has access to all institutions. When I send it from a user that has access just for his institution, I can't see the labels and the inventory anymore when the receiver confirms the transfer.

Comment (omelao) [reply app/controllers/boxes_controller.rb:29]: The sender should print this, even when the box institution is the receiver's institution.

File: app/controllers/boxes_controller.rb
Comment (straight-shoota) [inline app/controllers/boxes_controller.rb:2]:
```suggestion
  before_action :load_box, only: %i[show destroy]
```
No need for `delete`, that was just a typo from me.

**Pull request merged**
